xdsl/dialects/builtin.py

```diff
@@ -53,12 +53,12 @@ def __init__(self: ArrayAttr[_ArrayAttrT],
     @staticmethod
     def parse_parameter(parser: BaseParser) -> tuple[_ArrayAttrT]:
         parser.parse_char("[")
-        data: list[_ArrayAttrT] = parser.parse_list(
-            parser.parse_optional_attribute)
+        data = parser.parse_list_of(parser.try_parse_attribute,
+                                    "Expected attribute")
```

Comment:
Maybe a better error message? Something along the lines of `Malformed ArrayAttr: Expected an attribute here`.

I think it's valuable to have very greppable error messages (I know I'm also not perfect in that regard). Giving a little bit more context than you think you need to give in error messages is usually a good thing.

Also we might want some guidelines for good parse error messages.

Reply:
Yes! Hopefully in the future, we should try to have automatically something in the lines of `Error in builtin.array attribute parsing: ...` so we won't have to add that ;)

Reply:
Hmm, it reads a bit like noise to me, I feel like error messages should be clear and concise. Errors with unique error codes are a great way to make them searchable also. But yeah, wrapping the terse parsing errors in a more verbose pretty wrapper when surfacing would be good.